### DISCUSSION TITLE
Nonmonotonicity with CFE_TIME_GetTime

### DISCUSSION THREAD

**Discussion by 2shaar2059** [nasa/cFS#897] [Q&A]
**Describe the bug**
The time returned by CFE_TIME_GetTime() is not monotonically increasing.

**To Reproduce**
Create a new app (add the appropriate configuration to startup script and targets.cmake) with the following files:

runner.c:
```C
#include "cfe.h"
#include "cfe_error.h"
#include "cfe_es.h"
#include "cfe_evs.h"
#include "cfe_sb.h"
#include "unistd.h"

void TIME_AppMain(void)
{
    uint32_t prevSeconds;
    uint32_t prevSubseconds;
    long     counter = 0;

    while (true)
    {
        CFE_TIME_SysTime_t currTimestamp = CFE_TIME_GetTime();

        uint32_t currSeconds    = currTimestamp.Seconds;
        uint32_t currSubseconds = currTimestamp.Subseconds;

        if (counter > 0)
        {
            if (prevSeconds > currSeconds || (prevSeconds == currSeconds && prevSubseconds > currSubseconds))
            {
                CFE_ES_WriteToSysLog("Timestamp Discrepancy");

                CFE_ES_WriteToSysLog("Previous Timestamp: Seconds: %u\tSubseconds: %u", prevSeconds, prevSubseconds);

                CFE_ES_WriteToSysLog("Current Timestamp: Seconds: %u\tSubseconds: %u", currSeconds, currSubseconds);

                CFE_ES_WriteToSysLog("timestamp count: %ld ", counter);
            }
        }
        usleep(1000);
        counter++;
        prevSeconds    = currSeconds;
        prevSubseconds = currSubseconds;
    }
}
```

analyze_time/CMakeLists.txt:
```cmake
project(CFE_ANALYZE_TIME C)

add_cfe_app(analyze_time runner.c)
```

**Expected behavior**
There should never be any case when a timestamp has the same Seconds but lower Subseconds field than an earlier timestamp.

**Code snips**
see above

**System observed on:**
 - OS: 5.4.0-100-generic Ubuntu

**Additional context**
The frequency with which the time discrepancy is printed decreases when the input to the usleep() function is increased.  Still, there is always one timestamp discrepancy printed regardless of the input to usleep(). We first noticed the behavior while trying to read data from an IMU sending values at roughly 125Hz.

**Comment by 2shaar2059** [comment]
Tagging @bkolligs

**Comment by jphickey** [comment]
I don't think the CFE_TIME_GetTime has any guarantee of being monotonic.  

This function calculates a time (which theoretically would correlate to either UTC/TAI depending on what is selected).

This calculation is all based on the concept of "Mission Elapsed Time" (MET) which in turn depends on the value of the "Spacecraft Time Correlation Factor" (SCTF) that expresses the difference between the MET and civil calendar.

Both these values can be updated via "time at tone" signals.

The implementation of CFE TIME is such that updates from the PSP will jump/step the clock, it does not attempt to slew the clock to soften the change.

If you want monotonic, use the `CFE_PSP_GetTime()` API, which should be normalized to standard time units.  

Also, not sure why this was written in the "sample_app" repo, as it doesn't seem related to sample_app at all.

**Comment by bkolligs** [comment]
Thanks for the reply Joseph, we will test out this function and post updates here.
